book.md

```diff
@@ -680,7 +679,255 @@ signal handler.  The ability to fake a signal handler frame and call
+case where an attacker has obtained arbitrary read and write primitives, we
+need something more.
+
+### CFI
+
+[Control-flow integrity
+(CFI)](https://en.wikipedia.org/wiki/Control-flow_integrity)\index{CFI} is a
+family of mitigations that aim to preserve the intended control flow of a
+program. This is done by restricting the possible targets of indirect branches
+and returns.  A scheme that protects indirect jumps and calls is referred to as
+forward-edge CFI, whereas a scheme that protects returns is said to implement
+backward-edge CFI. Ideally, a CFI scheme would not allow any control flow
+transfers that don't occur in a correct program execution, however different
+schemes have varying granularities. They often rely on function type checks or
+use static analysis (points-to analysis) to identify potential control flow
+transfer targets.  [@Burow2017] compares a number of available CFI schemes
+based on the precision. For forward-edge CFI schemes, for example, schemes are
+classified based on whether or not they perform, among others, flow-sensitive
+analysis, context-sensitive analysis and class-hierarchy analysis.
+
+### Clang CFI
+
+[Clang's CFI](https://clang.llvm.org/docs/ControlFlowIntegrity.html) includes a
+variety of forward-edge control-flow integrity checks. These include checking
+that the target of an indirect function call is an address-taken function of
+the correct type and checking that a C++ virtual call happens on an object of
+the correct dynamic type.
+
+For example, assume we have a class `A` with a virtual function `foo` and a
+class `B` deriving from `A`, and that these classes are not exported to
+other compilation modules:
+
+```
+class A {
+public:
+  virtual void foo() {}
+};
+
+class B : public A {
+public:
+  virtual void foo() {}
+};
+
+void call_foo(A* a) {
+  a->foo();
+}
+```
+
+When compiling with `-fsanitize=cfi -flto -fvisibility=hidden` [^cfi-flags],
+the code for `call_foo` would look something like this:
+
+```
+<call_foo(A*)>:
+  stp     x29, x30, [sp, #-16]!
+  mov     x29, sp
+  ldr     x8, [x0]                   // load vptr
+  adrp    x9, 400000 <_init-0x558>
+  add     x9, x9, #0x858
+  sub     x9, x8, x9
+  sub     x9, x9, #0x10
+  ror     x9, x9, #5
+  cmp     x9, #0x2
+  b.cs    4006ec <call_foo(A*)+0x38>
+  ldr     x8, [x8]
+  blr     x8
+  ldp     x29, x30, [sp], #16
+  ret
+  brk     #0x1
+```
+
+This code looks complicated, but what it does is check that the vptr of
+the argument points to the vtable of `A` or of `B`, which are stored
+consecutively and are the only allowed possibilities. The checks generated for
+different types of control-flow transfers are similar.
+
+[^cfi-flags]: The LTO and visibility flags are required by Clang's CFI.
+
+Another implementation of forward-edge CFI is Windows [Control Flow
+Guard](https://docs.microsoft.com/en-us/windows/win32/secbp/control-flow-guard),
+which only allows indirect calls to functions that are valid indirect
+control flow targets.
```

Comment:
I wonder which functions are valid indirect control flow targets. Is it exactly the functions whose addresses are taken explicitly in source? Or something different? Or is that too complex to explain briefly here?

Reply:
I don't think it's only address-taken functions, AFAIK this also includes exported library functions (based on https://connormcgarr.github.io/examining-xfg/). It probably includes C++ virtual functions too. But since this isn't described in the official documentation, I intentionally left it a bit vague. I think I should change this to say "are marked as valid control flow targets" rather than "are valid ...", though.